extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/HybridJacksonPool.java

```diff
@@ -0,0 +1,141 @@
+package io.quarkus.jackson.runtime;
+
+import com.fasterxml.jackson.core.util.BufferRecycler;
+import com.fasterxml.jackson.core.util.BufferRecyclerPool;
+import org.jctools.queues.MpmcUnboundedXaddArrayQueue;
+import org.jctools.util.Pow2;
+import org.jctools.util.UnsafeAccess;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+
+public class HybridJacksonPool implements BufferRecyclerPool {
+
+    static final BufferRecyclerPool INSTANCE = new HybridJacksonPool();
+
+    private HybridJacksonPool() { }
+
+    private static final Predicate<Thread> isVirtual = findIsVirtual();
+
+    private static Predicate<Thread> findIsVirtual() {
+        try {
+            MethodHandle virtualMh = MethodHandles.publicLookup().findVirtual(Thread.class, "isVirtual", MethodType.methodType(boolean.class));
+            return t -> {
+                try {
+                    return (boolean) virtualMh.invokeExact(t);
+                } catch (Throwable e) {
+                    throw new RuntimeException(e);
+                }
+            };
+        } catch (Exception e) {
+            return t -> false;
+        }
+    }
+
+    private final BufferRecyclerPool nativePool = BufferRecyclerPool.threadLocalPool();
+
+    static class VirtualPoolHolder {
+        // Lazy on-demand initialization
+        private static final BufferRecyclerPool virtualPool = new StripedJCToolsPool(4);
+    }
+
+    @Override
+    public BufferRecycler acquireBufferRecycler() {
+        return isVirtual.test(Thread.currentThread()) ?
+                VirtualPoolHolder.virtualPool.acquireBufferRecycler() :
+                nativePool.acquireBufferRecycler();
+    }
+
+    @Override
+    public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+        if (bufferRecycler instanceof VThreadBufferRecycler) {
+            // if it is a PooledBufferRecycler it has been acquired by a virtual thread, so it has to be release to the same pool
+            VirtualPoolHolder.virtualPool.releaseBufferRecycler(bufferRecycler);
+        }
+        // the native thread pool is based on ThreadLocal, so it doesn't have anything to do on release
+    }
+
+    private static class StripedJCToolsPool implements BufferRecyclerPool {
+
+        private static final long PROBE = getProbeOffset();
+
+        private final int mask;
+
+        private final MpmcUnboundedXaddArrayQueue<BufferRecycler>[] queues;
```

Comment:
This queue is cool, but uses `Unsafe` and I didn't (ehm ehm my fault!) implemented a varhandle/atomicref/long version of it, as per other JCTools variants...so maybe @cescoffier got some suggestion about native image or the approach to follow here; I'm not aware of quarkus code directly using Unsafe, but I can be wrong (I didn't searched it)...

Last but not least: JCTOols is great (not by merit :P), but the benchmarks doesn't tell it to be the best *by large margin* tool for the job, because of how cpu worksl actually, in short:

- the mpmc queue uses an array adding/removing elements in FIFO order out of it, BUT: if the queue is empty or with less than 16/32 elements in, producer(s) and consumer(s) will perform store(s) (by adding new values, or reading and adding `null` to mark consumption) in the same cache-line, competing for it. The Cache Coherency protocol of CPU will make the communication of all changes around it the primary bottleneck of it. In order to benefit from using a queue (and get the contention to spread between head/tail), you need a much larger already populated pool (more than 32 elements), but it would cost! In the "real" world (but we can add telemetry and report it, to see if I'm right or not!) the queue will be mostly empty, causing this contention to happen for real, slowing down everyone
- there is a LOCK_FREE variant which use a single atomic ref as the top of a linked-list stack: i'm not a BIG FAN of linked-lists, but still, for the most practical cases,it will deliver similar performance, no Unsafe nor JCTools usage, and it's all jackson stuff

Reply:
> I'm not aware of quarkus code directly using Unsafe

Directly, no, we do not use it, but JBoss Threads which we do lean on uses it and I don't see any GraalVM substitutions for it, so I assume it works in native

Reply:
>  no Unsafe nor JCTools usage, and it's all jackson stuff

All else being equal, this would definitely be preferable for Quarkus

Reply:
> > no Unsafe nor JCTools usage, and it's all jackson stuff
> 
> All else being equal, this would definitely be preferable for Quarkus

I agree, I discussed this yesterday with @franz1981 and we believe that we can achieve similar performances without using Unsafe or JCTools. I will work on this in the next days.